src/main.rs

```diff
@@ -1,3 +1,72 @@
+extern crate pancurses;
+
+use pancurses::Window;
+use pancurses::{initscr, endwin, Input, noecho};
+
 fn main() {
-    println!("Hello, world!");
+    let window = initscr();
+    window.refresh();
+    window.keypad(true);
+    noecho();
+    render_line((0, 0), (40, 10), &window);
+    loop {
+        match window.getch() {
+            Some(Input::Character('q')) => break,
+            _ => ()
+        }
+    }
+    endwin();
+}
+
+trait RenderOutput {
+    fn set_pixel(&self, x: i32, y: i32);
+    fn get_size(&self) -> (i32, i32);
+}
+
+impl RenderOutput for Window {
+    fn set_pixel(&self, x: i32, y: i32) {
+        self.mvaddch(y, x, '#');
```

Comment:
I thought that this would yell at me to make `self` a mutable reference, since (presumably) `mvaddch()` is changing some internal state of the Window.  Alas, it compiles as is.  Any idea why I don't need to make it mutable?

Reply:
## High Level Answer

This comes back to the maybe-poor naming of `mut` which is more like "mutually exclusive" or something instead of "mutable" (although everyone says mutable and it _is_ short for mutable but it's unfortunately confusing in times like these :sweat_smile:). You only need a `&mut` to mutate something if _the method of mutation is unsynchronized_.

## Lower Level Answer

### Data Races

In general Rust is aimed at preventing undefined behavior. One type of undefined behavior is a data race (N.B. this is _not_ the same as a [_race condition_](https://doc.rust-lang.org/nomicon/races.html)!). A data race is when multiple operations occur concurrently on the same piece of memory, at least one is a write, and at least one is unsynchronized. For example Rust needs to stop you from mutating a `String` from two threads. Mutating a `String` is unsynchronized and so we'd have a data race. Rust prevents this by requiring a `&mut` for mutation which There Can Only Be One of.

Does that mean we can't modify _anything_ concurrently?

### Synchronized Data Structures

Of course not! Wrapping a data structure in a `Mutex`, for instance, synchronizes the write operation which removes the data race:

```rust
use std::sync::{Arc, Mutex};
fn change_string(s: Arc<Mutex<String>>) {
    let s1 = s.clone();
    let s2 = s.clone();
    std::thread::spawn(move || s1.lock().unwrap().push_str("foo"));
    std::thread::spawn(move || s2.lock().unwrap().push_str("bar"));
}
```

Here the `Mutex` ensures writes are synchronized so it exposes a mechanism for getting a mutable reference from a non-mutable borrow. Here neither `s`, `s1`, nor `s2` are marked as `mut` because writes are synchronized and cannot cause a data race! (To get a `&mut` you have to call `lock()` which `unsafe`ly handles the synchronization!)

Are locks the only way?

### Atomics

Negative! The most common, lockless synchronized data structures in the standard library are the atomics. Modifications to an atomic variable are, by the OS, implicitly synchronized so the following is A-OK:

```rust
use std::sync::Arc;
use std::sync::atomic::{AtomicU8, Ordering};
fn change_n(n: Arc<AtomicU8>) {
    let n1 = n.clone();
    let n2 = n.clone();
    std::thread::spawn(move || n1.fetch_add(1, Ordering::Relaxed));
    std::thread::spawn(move || n2.fetch_add(2, Ordering::Relaxed));
}
```

This is pretty darn confusing! It means that you can mutate "non-mutable" custom data structures like this:

```rust
use std::sync::Arc;
use std::sync::atomic::{AtomicU8, Ordering};

struct MyThing { inner: AtomicU8 }
impl MyThing {
    fn modify_me(&self) {
        self.inner.fetch_add(1, Ordering::Relaxed);
    }
}
```

and the reason is because this _cannot cause undefined behavior_ (which is really all Rust wants to stop us from) because the write is synchronized.

## What Is `mvaddch` Doing?

No idea! [The source](https://docs.rs/pancurses/0.16.1/src/pancurses/window.rs.html#385) calls out to [`ncurses`](https://github.com/ihalila/pancurses/blob/master/src/lib.rs#L9) which calls some [C function](https://docs.rs/ncurses/5.101.0/src/ncurses/ll.rs.html#135) which is, apparently, safe :sweat_smile: Either way, they've chosen to expose a function that _hides the synchronization_ much like `modify_me` above.